HARK/ConsumptionSaving/ConsRiskyAssetModel.py

```diff
@@ -0,0 +1,1944 @@
+        # A marginal unit of n will be withdrawn and put into m. Then consumed.
+        dvdnFuncRebAdj_term = lambda m,n: dvdmFuncRebAdj_term(m,n)/(1+tau)
+        
+        RebStageSol = RiskyContribRebSolution(
+            # Rebalancing stage
+            vFuncRebAdj = vFuncRebAdj_term,
+            DFuncAdj = DFuncAdj_term,
+            dvdmFuncRebAdj = dvdmFuncRebAdj_term,
+            dvdnFuncRebAdj = dvdnFuncRebAdj_term,
+            
+            # Adjusting stage
+            vFuncRebFxd = vFuncCns_term,
+            DFuncFxd = ConstantFunction(0.0),
+            dvdmFuncRebFxd = dvdmFuncCns_term,
+            dvdnFuncRebFxd = dvdnFuncCns_term,
+            dvdsFuncRebFxd = dvdsFuncCns_term)
+        
+        # Construct the terminal period solution
+        self.solution_terminal = RiskyContribSolution(RebStageSol,
+                                                      ShaStageSol,
+                                                      CnsStageSol)
+        
+    
+    def updateTau(self):
+        
+        if (type(self.tau) is list and (len(self.tau) == self.T_cycle)):
+            self.addToTimeVary('tau')
+        elif type(self.tau) is list:
+            raise AttributeError('If tau is time-varying, it must have length of T_cycle!')
+        else:
+            self.addToTimeInv('tau')
+        
+    def updateRiskyDstn(self):
+        '''
+        Creates the attributes RiskyDstn from the primitive attributes RiskyAvg,
+        RiskyStd, and RiskyCount, approximating the (perceived) distribution of
+        returns in each period of the cycle.
+        
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        None
+        '''
+        # Determine whether this instance has time-varying risk perceptions
+        if (type(self.RiskyAvg) is list) and (type(self.RiskyStd) is list) and (len(self.RiskyAvg) == len(self.RiskyStd)) and (len(self.RiskyAvg) == self.T_cycle):
+            self.addToTimeVary('RiskyAvg','RiskyStd')
+        elif (type(self.RiskyStd) is list) or (type(self.RiskyAvg) is list):
+            raise AttributeError('If RiskyAvg is time-varying, then RiskyStd must be as well, and they must both have length of T_cycle!')
+        else:
+            self.addToTimeInv('RiskyAvg','RiskyStd')
+        
+        # Generate a discrete approximation to the risky return distribution if the
+        # agent has age-varying beliefs about the risky asset
+        if 'RiskyAvg' in self.time_vary:
+            RiskyDstn = []
+            for t in range(self.T_cycle):
+                RiskyAvgSqrd = self.RiskyAvg[t] ** 2
+                RiskyVar = self.RiskyStd[t] ** 2
+                mu = np.log(self.RiskyAvg[t] / (np.sqrt(1. + RiskyVar / RiskyAvgSqrd)))
+                sigma = np.sqrt(np.log(1. + RiskyVar / RiskyAvgSqrd))
+                RiskyDstn.append(Lognormal(mu=mu, sigma=sigma).approx(self.RiskyCount))
+            self.RiskyDstn = RiskyDstn
```

Comment:
For a routine mathematical operation like this, it would be much better if it were packaged as a method where it can be reused.

Please consider implementing the more general solution as a part of the `distributions` module, as per this ticket:

https://github.com/econ-ark/HARK/issues/637

That would reduce the lines of code in the already quite complicated solution code.

Reply:
I grabbed this from `ConsPortfolioModel`. Yes, I think it would work better as a tool in `distributions`.